demos/Progress Bar/main.py

```diff
@@ -0,0 +1,73 @@
+import gi
+
+gi.require_version("Adw", "1")
+from gi.repository import Adw, GLib
+import workbench
+
+first_bar = workbench.builder.get_object("first")
+second_bar = workbench.builder.get_object("second")
+play = workbench.builder.get_object("play")
+progress_tracker = workbench.builder.get_object("progress_tracker")
+
+target = Adw.PropertyAnimationTarget.new(first_bar, "fraction")
+
+animation = Adw.TimedAnimation(
+    widget=first_bar,
+    value_from=0.2,
+    value_to=1,
+    duration=11000,
+    easing=Adw.Easing.LINEAR,
+    target=target,
+)
+
+
+def on_play_clicked(_button):
+    animation.play()
+    update_tracker()
+    pulse_progress()
+
+
+animation.connect("done", lambda _: animation.reset())
+
+play.connect("clicked", on_play_clicked)
+
+
+def on_pulse():
+    global counter
+    if counter >= 1.0:
+        counter = 0
+        second_bar.set_fraction(0)
+        return False
+
+    second_bar.pulse()
+    counter += increment
+    return True
+
+
+def pulse_progress():
```

Comment:
Please restructure the code in a way that does not require global variables, since the JS demo also doesn't need them. You can define functions within functions, those work as closures in Python.

Reply:
You mean using nonlocal variables like in the fixup commit?